src/credentials/CHIPCert.h

```diff
@@ -0,0 +1,460 @@
+    uint16_t EncodedCertLen;
+    uint16_t CertFlags;
+    uint16_t KeyUsageFlags;
+    uint16_t PubKeyAlgoOID;
+    uint16_t SigAlgoOID;
+    uint8_t CertType;
+    uint8_t KeyPurposeFlags;
+    uint16_t NotBeforeDate;
+    uint16_t NotAfterDate;
+    uint8_t PathLenConstraint;
+    uint8_t TBSHash[chip::Crypto::kSHA256_Hash_Length];
+};
+
+/**
+ *  @class ValidationContext
+ *
+ *  @brief
+ *    Context information used during certification validation.
+ */
+class ValidationContext
+{
+public:
+    uint32_t EffectiveTime;
+    ChipCertificateData * TrustAnchor;
+    ChipCertificateData * SigningCert;
+    uint16_t RequiredKeyUsages;
+    uint16_t ValidateFlags;
+#if CHIP_CONFIG_DEBUG_CERT_VALIDATION
+    CHIP_ERROR * CertValidationResults;
+    uint8_t CertValidationResultsLen;
+#endif
+    uint8_t RequiredKeyPurposes;
+    uint8_t RequiredCertType;
+
+    void Reset();
+};
+
+/**
+ *  @class ChipCertificateSet
+ *
+ *  @brief
+ *    Collection of CHIP certificate data providing methods for
+ *    certificate validation and signature verification.
+ */
+class DLL_EXPORT ChipCertificateSet
+{
+public:
+    typedef void * (*AllocFunct)(size_t size);
+    typedef void (*FreeFunct)(void * p);
+
+    ChipCertificateSet(void);
+
+    ChipCertificateData * Certs; // [READ-ONLY] Pointer to array of certificate data
+    uint8_t CertCount;           // [READ-ONLY] Number of certificates in Certs array
+    uint8_t MaxCerts;            // [READ-ONLY] Length of Certs array
+
+    CHIP_ERROR Init(uint8_t maxCerts, uint16_t decodeBufSize);
+    CHIP_ERROR Init(uint8_t maxCerts, uint16_t decodeBufSize, AllocFunct allocFunct, FreeFunct freeFunct);
+    CHIP_ERROR Init(ChipCertificateData * certBuf, uint8_t certBufSize, uint8_t * decodeBuf, uint16_t decodeBufSize);
+    void Release(void);
+    void Clear(void);
+
+    CHIP_ERROR LoadCert(const uint8_t * chipCert, uint32_t chipCertLen, uint16_t decodeFlags, ChipCertificateData *& cert);
+    CHIP_ERROR LoadCert(TLVReader & reader, uint16_t decodeFlags, ChipCertificateData *& cert);
+    CHIP_ERROR LoadCerts(const uint8_t * chipCerts, uint32_t chipCertsLen, uint16_t decodeFlags);
+    CHIP_ERROR LoadCerts(TLVReader & reader, uint16_t decodeFlags);
+    CHIP_ERROR AddTrustedKey(uint64_t caId, OID curveOID, const uint8_t * pubKey, uint16_t pubKeyLen, const uint8_t * pubKeyId,
+                             uint16_t pubKeyIdLen);
+    ChipCertificateData * FindCert(const CertificateKeyId & subjectKeyId) const;
+    ChipCertificateData * LastCert(void) const { return (CertCount > 0) ? &Certs[CertCount - 1] : nullptr; }
+
+    CHIP_ERROR ValidateCert(ChipCertificateData & cert, ValidationContext & context);
+    CHIP_ERROR FindValidCert(const ChipDN & subjectDN, const CertificateKeyId & subjectKeyId, ValidationContext & context,
+                             ChipCertificateData *& cert);
+
+    CHIP_ERROR VerifySignature(ChipCertificateData & cert, ChipCertificateData & caCert);
+
+protected:
```

Comment:
Can these be private instead? Not sure if we ever want derived classes with direct access to these members.